[Flang][OpenMP][MLIR] Initial array section mapping MLIR -> LLVM-IR lowering utilising omp.bounds

agozillon opened the pull request:
This patch seeks to add initial lowering of OpenMP array sections within target region map clauses from MLIR to LLVM IR.

This patch seeks to support fixed sized contiguous (don't think OpenMP supports anything other than contiguous sections from my reading but i could be wrong) arrays initially, before looking toward assumed size and shaped arrays. The patch also currently does not include stride, it's left for future work. 

Although, assumed size works in some fashion (dummy arguments) with some minor alterations to the OMPEarlyOutliner, so it is possible changes made in the IsolatedFromAbove series may allow this to work with no further required patches.

It utilises the generated omp.bounds to calculate the size of the mapped OpenMP array (both for sectioned and un-sectioned arrays) as well as the offset to be passed to the kernel argument structure.

Alongside these changes some refactoring of how map data is handled is attempted, using a new MapData structure to keep track of information utilised in the lowering of mapped values.

The initial addition of a more complex createDeviceArgumentAccessor that utilises capture kinds similarly to (and loosely based on) Clang to generate different kernel argument accesses is also added.

A similar function for altering how the kernel argument is passed to the kernel argument structure on the host is also utilised (createAlteredByCaptureMap), which allows modification of the pointer/basePointer based on their capture (and bounds information). It's of note ByRef, is the default for explicit mappings and ByCopy will be the default for implicit captures, so the former is currently tested in this patch and the latter is not for the moment.

agozillon commented:
A large portion of this patch is unfortunately the addition of the MapData structure and the addition of utilising byref/bycopy but they're rather linked together. If there's a wish to split the patch please do suggest how you would like the split (e.g. what portions into seperate patches) done,

agozillon commented:
Small ping for some reviewer attention on this if at all possible please.

kiranchandramohan commented:
@TIFitis could you review? As this is changing code that you initially wrote and subsequently modified by @agozillon.

kiranchandramohan commented on mlir/test/Target/LLVMIR/omptarget-array-sectioning-host.mlir:
```diff
@@ -0,0 +1,80 @@
+// RUN: mlir-translate -mlir-to-llvmir %s | FileCheck %s
```
Could you minimize this test? May be run the MLIR CSE, DCE passes on it?

agozillon replied:
I can certainly try!

kiranchandramohan reviewed (commented):
> Alongside these changes some refactoring of how map data is handled is attempted, using a new MapData structure to keep track of information utilised in the lowering of mapped values.
> 
> The initial addition of a more complex createDeviceArgumentAccessor that utilises capture kinds similarly to (and loosely based on) Clang to generate different kernel argument accesses is also added.
> 
> A similar function for altering how the kernel argument is passed to the kernel argument structure on the host is also utilised (createAlteredByCaptureMap), which allows modification of the pointer/basePointer based on their capture (and bounds information). It's of note ByRef, is the default for explicit mappings and ByCopy will be the default for implicit captures, so the former is currently tested in this patch and the latter is not for the moment.

Would it be possible to push all the refactorings directly as an NFC patch? And then have the changes necessary for array-sections only in this patch?

TIFitis commented on mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:
```diff
@@ -1629,13 +1622,153 @@ getRefPtrIfDeclareTarget(mlir::Value value,
+// structures.
+// This function is somewhat equivalent to Clang's getExprTypeSize inside of
+// CGOpenMPRuntime.cpp.
+llvm::Value *getSizeInBytes(DataLayout &dl, const mlir::Type &type,
```
Consider the following example:
```
subroutine omp_target(a, b, c)
   integer, intent(in) :: a, b, c
   integer :: x(a, b, c)
   !$omp target map(tofrom : x)
   !$omp end target
end subroutine omp_target
```

Here's a slice of the llvm IR generated:
```
  %.offload_sizes = alloca [1 x i64], align 8
  %kernel_args = alloca %struct.__tgt_kernel_arguments, align 8
  %4 = load i32, ptr %0, align 4
  %5 = sext i32 %4 to i64
  %6 = icmp sgt i64 %5, 0
  %7 = select i1 %6, i64 %5, i64 0
  %8 = load i32, ptr %1, align 4
  %9 = sext i32 %8 to i64
  %10 = icmp sgt i64 %9, 0
  %11 = select i1 %10, i64 %9, i64 0
  %12 = load i32, ptr %2, align 4
  %13 = sext i32 %12 to i64
  %14 = icmp sgt i64 %13, 0
  %15 = select i1 %14, i64 %13, i64 0
  %16 = mul i64 1, %7
  %17 = mul i64 %16, %11
  %18 = mul i64 %17, %15
  %19 = alloca i32, i64 %18, align 4
  %20 = sub i64 %7, 1
  %21 = sub i64 %11, 1
  %22 = sub i64 %15, 1
  br label %entry

entry:                                            ; preds = %3
  %23 = sub i64 %20, 0
  %24 = add i64 %23, 1
  %25 = sub i64 %21, 0
  %26 = add i64 %25, 1
  %27 = mul i64 %24, %26
  %28 = sub i64 %22, 0
  %29 = add i64 %28, 1
  %30 = mul i64 %27, %29
  %31 = mul i64 %30, 4
  %34 = getelementptr inbounds [1 x i64], ptr %.offload_sizes, i32 0, i32 0
  store i64 %31, ptr %34, align 8
```

The above code basically tries to recompute `%30` which is the same as `%18` already present in the alloca instruction.

My view is that we don't need and should neither use nor generate a boundsOp unless explicit bounds have been provided by the user.

Others however, have already expressed that we would like to have a boundsOp present at all times and use it whenever possible as it favours a single solution for all cases. I am not strictly against this, but I prefer the former way of doing things.

And from what I can tell Clang also reuses `%18` for the `offload_size` here.

agozillon replied:
I'd prefer to always utilise the bounds as it makes calculating these trivial for all scenarios (fixed/ assumed shape / assumed size). Although, it is a shame we are re-computing the value in this case. I am fine with either method we choose though, so I'll leave it up to reviewers. There's a fairly length discussion relating to this topic here: https://github.com/llvm/llvm-project/pull/67164 

Please do let me know which direction the reviewers wish to go so the patch can progress!

kiranchandramohan replied:
I don't think we currently have a mechanism to get all the information about arrays represented as descriptors in the LLVM Dialect. Without such a mechanism, we cannot get all the information to do the mapping. `omp.bounds` provides all the information currently, so I think we should stick with that for uniformity.

agozillon replied:
I also prefer the uniformity and ease of use, but perhaps I'm to close to the source on this one. I do understand the dislike for the excess generated arguments though, and the possible performance impact if subsequent optimisation passes can't tidy it up. 

However, it's not the only thing I've encountered recently during the lowering that will need some polishing off when we can. Allocatables for some odd reason spawn multiple allocas of the descriptor + data structure that get the same value assigned to them, and then only one of them is actually used. Unsure why (perhaps someone else does and it has a good reason or maybe it's an OpenMP lowering oddity), but it is something that might be worth a look into in the future.

TIFitis commented on mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:
```diff
@@ -1629,13 +1622,153 @@ getRefPtrIfDeclareTarget(mlir::Value value,
+// llvm module for lowered mapped varibles or checking
+// if something is declare target (and retrieving the
+// value).
+struct MapData {
```
Most of the data in this struct seems to overlap with `llvm::OpenMPIRBuilder::MapInfoTy`.

Instead of creating a new structTy, I'd rather inherit the existing MapInfosTy and add to it.

Similar to `struct MapCombinedInfoTy : llvm::OpenMPIRBuilder::MapInfosTy` in `clang/lib/CodeGen/CGOpenMPRuntime.cpp`.

agozillon replied:
that's a good idea, I can certainly try to do so!

agozillon commented:
> > Alongside these changes some refactoring of how map data is handled is attempted, using a new MapData structure to keep track of information utilised in the lowering of mapped values.
> > The initial addition of a more complex createDeviceArgumentAccessor that utilises capture kinds similarly to (and loosely based on) Clang to generate different kernel argument accesses is also added.
> > A similar function for altering how the kernel argument is passed to the kernel argument structure on the host is also utilised (createAlteredByCaptureMap), which allows modification of the pointer/basePointer based on their capture (and bounds information). It's of note ByRef, is the default for explicit mappings and ByCopy will be the default for implicit captures, so the former is currently tested in this patch and the latter is not for the moment.
> 
> Would it be possible to push all the refactorings directly as an NFC patch? And then have the changes necessary for array-sections only in this patch?

I can try to do so.

agozillon commented:
Updated the PR to tidy up the test a little using the recommended optimisation passes and utilising the MapInfoTy from the OMPIRBuilder, thank you both for your suggestions. 

I will try to split the PR up on Monday or Tuesday next week, but feel free to further review this PR until they become available. 

The plan for splitting up the PR will likely be into 3 patches:
   1) NFC changes that are refactoring for the follow up additions 
   2) The addition of ByRef/ByCopy handling for values on the device/host border (marshalling data on host and accessing data 
        on the device) which begins to replicate Clang a bit more closely and allows a wider range of map behaviour
   3) The array sectioning specific lowering changes

There will likely be a dependency for these patches based on ordering e.g.  Patch 3 needs 2+1, 2 needs 1 and 1 will ideally be self contained. I am not sure based on the wording of the original request for splinting up comment @kiranchandramohan if you'd prefer if I just pushed the NFC changes directly without review (patch 1 in this case) or if you'd prefer it to go through review still?

TIFitis commented on mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:
```diff
@@ -1638,13 +1631,154 @@ getRefPtrIfDeclareTarget(mlir::Value value,
+// llvm module for lowered mapped varibles or checking
+// if something is declare target (and retrieving the
+// value) more than neccessary.
+struct MapData : llvm::OpenMPIRBuilder::MapInfosTy {
```
Maybe change the name to something similar like `CombinedMapInfo `or `MapInfoData `for uniformity.

agozillon replied:
Sure, I'll change it to MapInfoData

TIFitis commented on mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:
```diff
@@ -1638,13 +1631,154 @@ getRefPtrIfDeclareTarget(mlir::Value value,
+  // Stripped off array/pointer to get the underlying
+  // element type
+  llvm::SmallVector<llvm::Type *, 4> BaseType;
+};
```
The `append `method needs to be overloaded to append these new entries.

agozillon replied:
Will do

TIFitis commented on mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:
```diff
@@ -1638,13 +1631,154 @@ getRefPtrIfDeclareTarget(mlir::Value value,
+struct MapData : llvm::OpenMPIRBuilder::MapInfosTy {
+  llvm::SmallVector<bool, 4> IsDeclareTarget;
+  llvm::SmallVector<mlir::Operation *, 4> MapClause;
+  llvm::SmallVector<llvm::Value *, 4> OriginalValue;
```
Looks to me like `OriginalValue` holds same value as `Pointers`. If so, can we remove this?

agozillon replied:
It doesn't in a few cases (https://github.com/llvm/llvm-project/pull/68689/files#diff-2cbb5651f4570d81d55ac4198deda0f6f7341b2503479752ef2295da3774c586R2265), OriginalValue should never change, the idea is that it's the original unaltered value pointed to by the captured MLIR map_info. The BasePointer/Pointers can be reassigned to new instructions such as GEP offsets, declare target globals, etc. and align with the BasePointer/Pointers that will be assigned to the kernel argument structure

kiranchandramohan replied:
Probably worth adding as a comment.

kiranchandramohan replied:
@agozillon please add this as a comment.

kiranchandramohan commented:
@agozillon NFC changes can go directly in without a review. 

If @TIFitis is OK with the current changes in this patch and can approve then this can go in without breaking up into patches.

TIFitis commented:
> If @TIFitis is OK with the current changes in this patch and can approve then this can go in without breaking up into patches.

Once patch is updated, I'll approve it.

agozillon commented:
> @agozillon NFC changes can go directly in without a review.

Ah, I wasn't aware of that, thank you @kiranchandramohan for teaching me!

> If @TIFitis is OK with the current changes in this patch and can approve then this can go in without breaking up into patches.

Thank you very much.

agozillon commented:
most recent update is a rebase and attempt at addressing the review comments from @TIFitis via the implementation of an overloaded append and a more appropriate name for MapData (now MapInfoData)

TIFitis reviewed (approved):
LGTM 👍🏽

agozillon commented:
> LGTM 👍🏽

Thank you very much for the review @TIFitis, greatly appreciated!

@kiranchandramohan I'd love a final acceptance from you if at all possible.

I will land the patch tomorrow afternoon after another rebase, if no one has any further comments they wish to add.

kiranchandramohan commented on mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:
```diff
@@ -1642,13 +1635,165 @@ getRefPtrIfDeclareTarget(mlir::Value value,
+    if (auto arrTy = llvm::dyn_cast_if_present<LLVM::LLVMArrayType>(
+            ptrTy.getElementType())) {
+      underlyingTypeSzInBits = getArrayElementSizeInBits(arrTy, dl);
+    }
```
Should else `StructType` be a TODO?

agozillon replied:
In this case it's actually covered by the default case, it will calculate the full structure size in bits. We primarily just need the `if` to get the element type size to calculate the array size when using bounds. 

If we ever wish to drop the default bounds generation for arrays, we'd have to alter this segment slightly to get full array size calculation, but that wouldn't be too difficult.

kiranchandramohan commented on mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:
```diff
@@ -1642,13 +1635,165 @@ getRefPtrIfDeclareTarget(mlir::Value value,
+      for (auto bounds : memberClause.getBounds()) {
+        if (auto boundOp = mlir::dyn_cast_if_present<mlir::omp::DataBoundsOp>(
+                bounds.getDefiningOp())) {
+
```
A short comment regarding this computation something like `(elemCount*[UB-LB]+1)`?

kiranchandramohan commented on mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:
```diff
@@ -1642,13 +1635,165 @@ getRefPtrIfDeclareTarget(mlir::Value value,
+        if (auto boundOp = mlir::dyn_cast_if_present<mlir::omp::DataBoundsOp>(
+                bounds.getDefiningOp())) {
+
+          if (!elementCount) {
+            elementCount = builder.CreateAdd(
+                builder.CreateSub(
+                    moduleTranslation.lookupValue(boundOp.getUpperBound()),
+                    moduleTranslation.lookupValue(boundOp.getLowerBound())),
+                builder.getInt64(1));
+          } else {
+            elementCount = builder.CreateMul(
+                elementCount,
+                builder.CreateAdd(
+                    builder.CreateSub(
+                        moduleTranslation.lookupValue(boundOp.getUpperBound()),
+                        moduleTranslation.lookupValue(boundOp.getLowerBound())),
+                    builder.getInt64(1)));
+          }
+        }
```
Can you merge this by setting elemCount to 1 initially?

Or move the (UB-LB)+1 calculation outside the if-else.

kiranchandramohan commented on mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:
```diff
@@ -1642,13 +1635,165 @@ getRefPtrIfDeclareTarget(mlir::Value value,
+  for (mlir::Value mapValue : mapOperands) {
+    if (auto mapOp = mlir::dyn_cast_if_present<mlir::omp::MapInfoOp>(
+            mapValue.getDefiningOp())) {
+      if (llvm::Value *refPtr = getRefPtrIfDeclareTarget(
+              mapOp.getVarPtr(), moduleTranslation)) { // declare target
+        mapData.OriginalValue.push_back(
+            moduleTranslation.lookupValue(mapOp.getVarPtr()));
+        mapData.IsDeclareTarget.push_back(true);
+        mapData.BasePointers.push_back(refPtr);
+        mapData.Pointers.push_back(mapData.OriginalValue.back());
+      } else { // regular mapped variable
+        mapData.OriginalValue.push_back(
+            moduleTranslation.lookupValue(mapOp.getVarPtr()));
+        mapData.IsDeclareTarget.push_back(false);
+        mapData.BasePointers.push_back(mapData.OriginalValue.back());
+        mapData.Pointers.push_back(mapData.OriginalValue.back());
+      }
```
Hoist out the common values.

kiranchandramohan reviewed (commented):
I will need more time to go through this. Have left a few comments.

agozillon commented on mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:
```diff
@@ -2001,61 +2111,204 @@ static bool targetOpSupported(Operation &opInst) {
         userVec.push_back(user);
 
-      for (llvm::User *user : userVec) {
+      for (auto *user : userVec) {
```
think this might have just slipped in during some change set, thank you.

agozillon commented:
rebased and attempted to resolve the latest reviewer comments

agozillon commented:
> I will need more time to go through this. Have left a few comments.

No problem, take your time.

kiranchandramohan commented on openmp/libomptarget/test/offloading/fortran/basic-target-region-3D-array-section.f90:
```diff
@@ -0,0 +1,39 @@
+       end do
+    end do
+
+!$omp target map(tofrom:inArray(1:3, 1:3, 2:2), outArray(1:3, 1:3, 1:3), j, k)
```
Does this test need all of `outArray` to be mapped?

agozillon replied:
No, I simply chose to in this case to test a variety of different mappings didn't cause issues.

kiranchandramohan commented on mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:
```diff
@@ -2001,61 +2100,204 @@ static bool targetOpSupported(Operation &opInst) {
+          }
+        }
+      } break;
+      case mlir::omp::VariableCaptureKind::ByCopy: {
```
Is the `ByCopy` case tested?

agozillon replied:
Not upstream currently, downstream yes. It unfortunately isn't testable until @TIFitis IsolatedFromAbove patch series lands, which also incorporates implicit argument capture which for scalars defaults to ByCopy. At the moment explicit's all default to capture ByRef.

I could perhaps write an MLIR -> LLVM-IR test showcasing the transformation though, as opposed to a Fortran OpenMP -> executable test,  if that sounds reasonable in the interim?

kiranchandramohan replied:
In this location, we only have `MLIR -> LLVM-IR ` tests, so Fortran OpenMP -> executable test is not needed. You could alternatively move this code out of this patch into a separate patch.

agozillon replied:
I'd prefer to add an MLIR -> LLVM-IR patch for it and maintain the change in this patch if at all possible?

kiranchandramohan replied:
That is fine.

kiranchandramohan commented on mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:
```diff
@@ -2001,61 +2100,204 @@ static bool targetOpSupported(Operation &opInst) {
   }
 }
 
+// This currently implements a very light version of Clang's
+// EmitParmDecl's handling of direct argument handling as well
+// as a portion of the argument access generation based on
+// capture types found at the end of emitOutlinedFunctionPrologue
+// in Clang. The indirect path handling of EmitParmDecl's may be
+// required for future work, but a direct 1-to-1 copy doesn't seem
+// possible as the logic is rather scattered throughout Clang's
+// lowering and perhaps we wish to deviate slightly.
```
While the source of these changes are welcome, It would be more ideal to describe what exactly this function does.
What is `argument access generation`? What is `retval` here?

agozillon replied:
Happy to add an extension to the comment on this. 

The function generates different instructions for retrieving the kernel argument inside of the kernel based on capture type, this can change the semantics of the kernel such as creating an intermediate copy that won't be written back. It's currently a slight extension to the previous copyInput function that utilises ByRef/ByCopy to generate instructions more closely aligned to the kernels in Clang when faced with certain mapped data.

kiranchandramohan commented on mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:
```diff
@@ -2001,61 +2100,204 @@ static bool targetOpSupported(Operation &opInst) {
+    if (mapData.OriginalValue[i] == input) {
+      if (auto mapOp = mlir::dyn_cast_if_present<mlir::omp::MapInfoOp>(
+              mapData.MapClause[i])) {
+        capture = mapOp.getMapCaptureType().value_or(
```
Do you really mean to override the capture value here?

agozillon replied:
Yep, the previous assignment on line 2157 is providing a default, the value_or call is also providing a fall back if for some reason there is no provided capture type (should always be the case if generated by Fortran, but may not be if handwritten or generated by some other frontend/path). 

The loop is checking if we find an associated map clause with the input and if we do, we retrieve the capture type from it. Can be used for other things as well, such as finding the underlying type beneath the opaque pointer which may be necessary in certain cases.

kiranchandramohan replied:
Ok. And this is specifically doing this for `input`. This function will be easier to read with a description based on the following three inputs/outpts.
```
llvm::Argument &arg,
llvm::Value *input
llvm::Value *&retVal
```

agozillon replied:
Happy to expand the original comment for the function, with some argument related comments similar to the way the OMPIRBuilder notates function argument

kiranchandramohan replied:
The following arguments are obvious here. The other arguments are interesting and hence a description with the above arguments is possibly more valuable here.
```
                             llvm::OpenMPIRBuilder &ompBuilder,
                             LLVM::ModuleTranslation &moduleTranslation,
                             llvm::IRBuilderBase::InsertPoint allocaIP,
                             llvm::IRBuilderBase::InsertPoint codeGenIP
                             ```

kiranchandramohan reviewed (commented):
A few more comments.

agozillon commented:
Rebased and tried to address @kiranchandramohan's comments added two new mlir tests for bycopy vs byref lowering for host and device. And expanded on the comment for createDeviceArgumentAccessor. 

I haven't changed the bounds sizing for the 3D Array sectioning test as I wasn't too sure if you were just asking for clarity on it or not! I am happy to change it though if desired.

kiranchandramohan reviewed (commented):
Looks fine except for the `getElementType`.

agozillon commented on mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:
```diff
@@ -1642,13 +1635,154 @@ getRefPtrIfDeclareTarget(mlir::Value value,
+  llvm::Type *type = moduleTranslation.convertType(inputType);
+  if (auto pointerType =
+          llvm::dyn_cast<mlir::omp::PointerLikeType>(inputType)) {
+    if (auto eleType = pointerType.getElementType()) {
```
It won't stop working (currently is rebased on all opaque changes), but this is redundant at the moment, and with nested pointers it might lead to an assert, unsure though, but I'll see if I can remove the redundant segment!  

getLLVMIRType is already called with VarType as an argument to inputType at line 1768.

agozillon commented on mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:
```diff
@@ -1642,13 +1635,154 @@ getRefPtrIfDeclareTarget(mlir::Value value,
+                                   DataLayout &dl,
+                                   llvm::IRBuilderBase &builder) {
+  for (mlir::Value mapValue : mapOperands) {
+    if (auto mapOp = mlir::dyn_cast_if_present<mlir::omp::MapInfoOp>(
```
Should be, will do so.

kiranchandramohan reviewed (approved):
Thanks @agozillon for the changes. I missed one point, I recently converted all OpenMP translation tests to use opaque pointers. You will have to do that for all the tests before submitting, i.e remove the element-type from the !llvm.ptr.

LGTM.

agozillon commented:
> Thanks @agozillon for the changes. I missed one point, I recently converted all OpenMP translation tests to use opaque pointers. You will have to do that for all the tests before submitting, i.e remove the element-type from the !llvm.ptr.
> 
> LGTM.

Thank you very much @kiranchandramohan I'll upstream on Monday afternoon, give a small amount of leeway for other people to have more input.

I'll fix the tests before upstreaming, I think I modified a few on rebase, but the newer ones do indeed need to be fixed to the new opaque pointers, thank you for reminding me.

agozillon committed:
[Flang][OpenMP][MLIR] Initial array section mapping MLIR -> LLVM-IR lowering utilising omp.bounds

This patch seeks to add initial lowering of OpenMP array
sections within target region map clauses from MLIR to
LLVM IR.

This patch seeks to support fixed sized arrays initially,
before looking toward assumed size and shaped arrays.

Although, assumed size works in some fashion (dummy
arguments) with some minor alterations to the OMPEarlyOutliner,
so it is possible changes made in the IsolatedFromAbove series
may allow this to work with no further required patches.

It utilises the generated omp.bounds to calculate the size
of the mapped OpenMP array (both for sectioned and
un-sectioned arrays) as well as the offset to be passed to
the kernel argument structure.

Alongside these changes some refactoring of how map data is
handled is attempted, using a new MapData structure to keep
track of information utilised in the lowering of mapped values.

The initial addition of a more complex
createDeviceArgumentAccessor that utilises capture
kinds similarly to (and loosely based on) Clang to
generate different kernel argument accesses is also
added.

A similar function for altering how the kernel argument
is passed to the kernel argument structure on the host
is also utilised (createAlteredByCaptureMap), which
allows modification of the pointer/basePointer based
on their capture (and bounds information). It's of note
ByRef, is the default for explicit mappings and ByCopy
will be the default for implicit captures, so the former
is currently tested in this patch and the latter is not
for the moment.

agozillon commented:
Rebased and removed final typed-pointers from the tests and did some final local checks, will merge when the final in PR tests complete. 

So please do speak up if anyone has any final comments they wish to make or any want to delay this PR.

agozillon commented:
seems to be some related gfortran regressions caused by the PR looking into them at the moment.

agozillon commented:
The gfortran regressions should be fixed by: https://github.com/llvm/llvm-project/commit/68c384676cf92289d823576f915ed296d209354f which re-introduces the very basic implicit capture we have currently that I a little too hastily removed in this PR. It appears while we don't use it in any of our tests, we do depend on the behavior in the regression test suite to some extent. So until the IFA series of patches lands, we must continue to keep this little bit of behavior to satisfy the regression tests, when the IFA patches from @TIFitis land this segment can likely be removed (can be done in the IFA patches themselves, or I can do so after the fact).  

Unfortunately, that means at the moment we occasionally (such as when using do loops) capture constants that we don't handle very well yet. Hopefully the IFA patch fixes this when it lands, but if it doesn't it's something I'm happy to look into as it's resolved downstream, so should just need a little nudge somewhere in the right direction. But in any case, I had to change some tests to utilise do while's with explicit arguments.
